src/ScheduleFunctions.cpp

```diff
@@ -496,6 +498,40 @@ Stmt build_produce(Function f, const Target &target) {
         Stmt check = AssertStmt::make(EQ::make(result, 0), error);
         check = LetStmt::make(result_name, e, check);
 
+        if (!cropped_buffers.empty()) {
+            // We need to clean up the temporary crops we made for the
+            // outputs in case any of them have device allocations.
+            vector<Expr> cleanup_args;
+
+            // Make a struct with the buffers and their uncropped parents
+            for (auto p : cropped_buffers) {
+                // The cropped buffer_t
+                cleanup_args.push_back(p.first);
+                // Its parent
+                cleanup_args.push_back(p.second);
+            }
+
+            if (cropped_buffers.size() > 1) {
+                // NULL-terminate it
+                cleanup_args.push_back(make_zero(type_of<struct halide_buffer_t *>()));
+            }
+
+            Expr cleanup_struct = Call::make(Handle(),
+                                             Call::make_struct,
+                                             cleanup_args,
+                                             Call::Intrinsic);
+
+            // We have to anticipate failures in the extern stage, so
+            // call this by registering a destructor before the extern
+            // call and triggering it afterwards using an Allocate node.
```

Comment:
register_destructor can only trigger on function exit, and usually only on failure - I want RAII over a specific scope. Allocate does that.